### PR TITLE
Adds a mixed-state simulator: part 3

### PR DESCRIPTION
**Pull request by ixfoduap**
**Context:**
This is the third and last pull request in a series of changes leading to the creation of a new built-in simulator for mixed-state quantum computations. The base branch for this PR is the same as in #807, which is currently under review

**Description of the Change:**

This PR adds two missing methods, `_apply_basis_state()` and `_apply_state_vector()`, which are required to apply the operations `BasisState` and `QubitStateVector` respectively. The resulting `default.mixed` device has also been added to device integration testing. 

With these changes, a fully-functioning version of the device should be complete. Potential improvements remain possible, which may be considered in the future.

### CODE REVIEW

File: tests/devices/test_default_mixed.py
Comment (josh146) [inline tests/devices/test_default_mixed.py:633]: Because of the if statements, these feel like two separate tests

**Review by josh146 — APPROVED** [review]
Awesome to have this in @ixfoduap! It looks great from my side, the only thing that was a bit hard to code review were the tests --- they were slightly opaque, in that the 'expected' result was created by calling an external function. Tests are the one place where I am in favour of code duplication 😆 

Apart from that, awesome that the device integration tests all pass!

File: .github/workflows/tests.yml
Comment (trbromley) [inline .github/workflows/tests.yml:79]: Nice! Do we also need to update the changelog?

File: tests/devices/test_default_mixed.py
Comment (trbromley) [reply tests/devices/test_default_mixed.py:442]: Is there an overarching motivation for setting `atol` and `rtol`?

File: tests/devices/test_default_mixed.py
Comment (trbromley) [inline tests/devices/test_default_mixed.py:461]:
```suggestion
        """Tests that different basis states are applied correctly when applied to a subset of wires"""
```

**Review by trbromley — COMMENTED** [review]
Thanks @ixfoduap, looking good!

File: pennylane/devices/default_mixed.py
Comment (trbromley) [inline pennylane/devices/default_mixed.py:308]: Is pytest wrong? :thinking:

**Review by trbromley — APPROVED** [review]

File: .github/workflows/tests.yml
Comment (josh146) [reply .github/workflows/tests.yml:79]: yes! Definitely

File: tests/devices/test_default_mixed.py
Comment (josh146) [reply tests/devices/test_default_mixed.py:442]: I guess it's partially convention, partially to give us explicit control of the tolerance via a central fixture in `conftest.py`.

File: pennylane/devices/default_mixed.py
Comment (ixfoduap) [reply pennylane/devices/default_mixed.py:280]: Yes, I think there are many potential changes/improvements we could add. I'd prefer to sketch them out and then decide which, if any, we want to implement. For example we could also look at:
- tensordot version of `apply_operation`
- speedups for specific operations, like Tom did for `default.qubit`

File: tests/devices/test_default_mixed.py
Comment (ixfoduap) [reply tests/devices/test_default_mixed.py:440]: So, `basis_state()` is an internal function I created within the tests files. It's just 
```python3
def basis_state(index, nr_wires):
    rho = np.zeros((2 ** nr_wires, 2 ** nr_wires), dtype=np.complex128)
    rho[index, index] = 1
    return rho
```
Since specific states are used *a lot* throughout the tests, I found it best to define these very simple functions, whose correctness is almost self-evident. 

Do you still think it's worth making the tests even more explicit?

File: tests/devices/test_default_mixed.py
Comment (ixfoduap) [reply tests/devices/test_default_mixed.py:442]: Same as in #807 : why is this? I thought one of the advantages of using `np.allclose()` is that we didn't have to specify the tolerance explicitly...

File: .github/CHANGELOG.md
Comment (trbromley) [inline .github/CHANGELOG.md:23]:
```suggestion
  ...     return qml.expval(qml.PauliZ(0))
```

File: .github/CHANGELOG.md
Comment (trbromley) [inline .github/CHANGELOG.md:11]: Do we want to advertise `dev.state`? I've always thought of it as a more "behind-the-scenes" sort of thing, and would be ok if this snippet just had the device definition.

File: .github/CHANGELOG.md
Comment (trbromley) [inline .github/CHANGELOG.md:24]: Printing the state at this point would give the same as above - it first needs to be called with `circuit([0, np.pi])`

File: .github/CHANGELOG.md
Comment (trbromley) [inline .github/CHANGELOG.md:27]: Could maybe choose some different values to get a more interesting expectation value?

File: .github/CHANGELOG.md
Comment (trbromley) [inline .github/CHANGELOG.md:22]:
```suggestion
  ...     qml.AmplitudeDamping(0.5, wires=0)
```

File: pennylane/beta/tapes/__init__.py
Comment (trbromley) [inline pennylane/beta/tapes/__init__.py:20]: Wonder why this is showing up :thinking:

Comment (trbromley) [reply pennylane/beta/tapes/__init__.py:20]: @ixfoduap I believe the extra files are showing up due to the merge-in of `master`. It should be ok, but one option is to revert that commit and then do `git merge mixed_simulator2`

File: tests/devices/test_default_mixed.py
Comment (josh146) [reply tests/devices/test_default_mixed.py:440]: Personally, I would favour duplicating the two lines

```python
    rho = np.zeros((2 ** nr_wires, 2 ** nr_wires), dtype=np.complex128)
    rho[index, index] = 1
```

within the tests (but curious to get others thoughts as well)

File: .github/CHANGELOG.md
Comment (josh146) [reply .github/CHANGELOG.md:11]: :+1: Especially if we plan on having `return qml.state()` work and return density matrices for `default.mixed`.

File: .github/CHANGELOG.md
Comment (josh146) [reply .github/CHANGELOG.md:24]: Perhaps a reason not to advertise `dev.state`?

File: tests/devices/test_default_mixed.py
Comment (ixfoduap) [reply tests/devices/test_default_mixed.py:468]: I did a quick count and there are a total of 41 tests that would need to be changed this way. I can do it, but is it worth it?

Comment (josh146) [reply tests/devices/test_default_mixed.py:463]: No, this change isn't required for my approval; instead it is more of a 'nitpick' or personal preference as to how I would liked to have read/written the tests.

So up to you! Depends how convinced you are by my arguments 🙂

**Comment by josh146** [comment]
Hey @ixfoduap, just checking the status of this PR. Is it ready to be merged in?

**Comment by ixfoduap** [comment]
@josh146 It's ready to be merged but it was blocked by some tests not passing. Will address that today

**Pull request merged**
